### PR TITLE
Mark .rmeta files as /SAFESEH on x86 Windows.

### PR DESCRIPTION
**Pull request by zetafunction**
Chrome links .rlibs with /WHOLEARCHIVE or -Wl,--whole-archive to prevent the linker from discarding static initializers. This works well, except on Windows x86, where lld complains:

  error: /safeseh: lib.rmeta is not compatible with SEH

The fix is simply to mark the .rmeta as SAFESEH aware. This is trivially true, since the metadata file does not contain any executable code.

### CODE REVIEW

**Comment by zetafunction** [comment]
Is there a good place to put a helper function for managing the `@feat.00` symbol? Right now, this is copied from https://github.com/rust-lang/rust/blob/b2d2184edea578109a48ec3d8decbee5948e8f35/compiler/rustc_codegen_ssa/src/back/link.rs#L1901 which I discovered in the course of fixing this bug. @nbdd0121 @petrochenkov

**Comment by bjorn3** [comment]
You can probably put it in `create_object_file` and then remove this code from link.rs.

**Comment by zetafunction** [comment]
> You can probably put it in `create_object_file` and then remove this code from link.rs.

Thanks, that's a good suggestion. I moved it to `create_object_file()`; however, I made it so `create_object_file()` always returns an object with mangling enabled.

`link.rs` separately (re-)disables mangling.

I think having `create_object_file()` silently return an Object with mangling disabled only for COFF objects is too action-at-a-distance.

File: tests/run-make/windows-safeseh/Makefile [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+foo:
+	$(RUSTC) -C linker=rust-lld foo.rs
+
+# Ensure that LLD can link when /WHOLEARCHIVE: is used with an .rlib.
+# Previously, lib.rmeta was not marked as (trivially) SAFESEH-aware.
+bar: baz
+	$(RUSTC) -C linker=rust-lld -C link-arg=/WHOLEARCHIVE:libbaz.rlib bar.rs
+
+baz:
+	$(RUSTC) baz.rs
```
Comment (bjorn3) [inline tests/run-make/windows-safeseh/Makefile:19]: nit: missing trailing newline

**Comment by bjorn3** [comment]
> Thanks, that's a good suggestion. I moved it to `create_object_file()`; however, I made it so `create_object_file()` always returns an object with mangling enabled.
> 
> `link.rs` separately (re-)disables mangling.
> 
> I think having `create_object_file()` silently return an Object with mangling disabled only for COFF objects is too action-at-a-distance.

Makes sense.

LGTM with the above nit fixed.

File: tests/run-make/windows-safeseh/Makefile (hunk 1 above)
Comment (zetafunction) [reply tests/run-make/windows-safeseh/Makefile]: Done. Sorry I didn't have my editor set up correctly.

**Pull request merged**